In [2]:
from pathlib import Path
import pymupdf
from IPython.display import Image, display

In [3]:
DATA_DIR = Path("../data/raw")
list(DATA_DIR.glob("*.pdf"))

[WindowsPath('../data/raw/microsoft_report.pdf'),
 WindowsPath('../data/raw/worldbank_report.pdf')]

In [4]:
pdf_path = next(DATA_DIR.glob("*world*")) # next() pour récupérer la valeur de l'objet créer par glob()

print(pdf_path)

..\data\raw\worldbank_report.pdf


In [5]:
doc = pymupdf.open(pdf_path)

print(f"Nombre de pages : {len(doc)}")
print(f"Métadonnées : {doc.metadata}")

Nombre de pages : 38
Métadonnées : {'format': 'PDF 1.4', 'title': '', 'author': '', 'subject': '', 'keywords': '', 'creator': 'Adobe InDesign 20.5 (Macintosh)', 'producer': 'Adobe PDF Library 17.0', 'creationDate': "D:20250908153744-06'00'", 'modDate': "D:20250924155840-04'00'", 'trapped': '', 'encryption': None}


In [6]:
page = doc[12]
print(page.get_text("text")[:2000])

22
CIRDI RAPPORT ANNUEL 2025
23
RÉPARTITION DES AFFAIRES ENREGISTRÉES AU COURS DE L’EXERCICE 2025 
SELON LE SECTEUR ÉCONOMIQUE
Au cours de l’exercice 2025, un total de 233 nominations ont été effectuées auprès de 
tribunaux, d’une commission de conciliation et de comités ad hoc CIRDI. Ces nominations 
ont été effectuées au sein de 51 tribunaux dans des arbitrages initiaux, au sein d’un tribunal 
dans le cadre d’une procédure de réexamen, au sein d’un tribunal dans le cadre d’une 
procédure de révision, et au sein d’une commission de conciliation. D’autres nominations 
ont été faites au sein de 22 comités ad hoc. En outre, il a fallu procéder à 5 nominations dans 
le cadre de la reconstitution de tribunaux. 
Soixante-six pour cent des nominations ont été effectuées soit par les parties, soit 
par les arbitres désignés par celles-ci, tandis que les 34% restant l’ont été par le CIRDI 
conformément à un accord entre les parties ou aux clauses applicables en cas de 
défaillance des parties.

## Block

In [7]:
# Texte sous forme de block (information spatiale, x0,y0,x1,y1,text,block_no,block_type)
blocks = page.get_text("blocks")
print(f"Nombre de blocks : {len(blocks)}")

for block in blocks[:5]:
    print(block)
    print("-" * 80)

Nombre de blocks : 25
(35.70000076293945, 712.8818969726562, 1116.0, 724.722900390625, '22\nCIRDI RAPPORT ANNUEL 2025\n23\n', 0, 0)
--------------------------------------------------------------------------------
(186.052001953125, 58.70428466796875, 516.6380615234375, 82.30428314208984, 'RÉPARTITION DES AFFAIRES ENREGISTRÉES AU COURS DE L’EXERCICE 2025 \nSELON LE SECTEUR ÉCONOMIQUE\n', 1, 0)
--------------------------------------------------------------------------------
(187.4694061279297, 448.2207946777344, 542.1524658203125, 531.0718383789062, 'Au cours de l’exercice 2025, un total de 233 nominations ont été effectuées auprès de \ntribunaux, d’une commission de conciliation et de comités ad hoc CIRDI. Ces nominations \nont été effectuées au sein de 51 tribunaux dans des arbitrages initiaux, au sein d’un tribunal \ndans le cadre d’une procédure de réexamen, au sein d’un tribunal dans le cadre d’une \nprocédure de révision, et au sein d’une commission de conciliation. D’autres nomina

Blocks sous forme de tuple

## Dict

In [8]:
# Représentation structurée et détaillée du contenu de la page
page = doc[12]

page_dict = page.get_text("dict")

In [9]:
print(page_dict)
print("\n")
print(type(page_dict))
print("\n")
print(page_dict.keys())

{'width': 1152.0, 'height': 756.0, 'blocks': [{'type': 0, 'number': 0, 'flags': 0, 'bbox': (35.70000076293945, 712.8818969726562, 1116.0, 724.722900390625), 'lines': [{'spans': [{'size': 9.0, 'flags': 20, 'bidi': 0, 'char_flags': 24, 'font': 'AndesBold', 'color': 2301728, 'alpha': 255, 'ascender': 0.9679999947547913, 'descender': -0.3409999907016754, 'text': '22', 'origin': (35.70000076293945, 721.6539306640625), 'bbox': (35.70000076293945, 712.9419555664062, 46.35599899291992, 724.722900390625)}], 'wmode': 0, 'dir': (1.0, 0.0), 'bbox': (35.70000076293945, 712.9419555664062, 46.35599899291992, 724.722900390625)}, {'spans': [{'size': 7.0, 'flags': 4, 'bidi': 0, 'char_flags': 16, 'font': 'AndesBook', 'color': 2301728, 'alpha': 255, 'ascender': 0.9480000138282776, 'descender': -0.257999986410141, 'text': 'CIRDI RAPPORT ANNUEL 2025', 'origin': (55.0, 720.8438720703125), 'bbox': (55.0, 714.2078857421875, 152.31399536132812, 722.64990234375)}], 'wmode': 0, 'dir': (1.0, 0.0), 'bbox': (55.0, 7

In [10]:
print("Largeur :", page_dict["width"])
print("Hauteur :", page_dict["height"])
print("Nombre de blocs :", len(page_dict["blocks"]))

Largeur : 1152.0
Hauteur : 756.0
Nombre de blocs : 25


In [11]:
block = page_dict["blocks"][0]
print(block)

{'type': 0, 'number': 0, 'flags': 0, 'bbox': (35.70000076293945, 712.8818969726562, 1116.0, 724.722900390625), 'lines': [{'spans': [{'size': 9.0, 'flags': 20, 'bidi': 0, 'char_flags': 24, 'font': 'AndesBold', 'color': 2301728, 'alpha': 255, 'ascender': 0.9679999947547913, 'descender': -0.3409999907016754, 'text': '22', 'origin': (35.70000076293945, 721.6539306640625), 'bbox': (35.70000076293945, 712.9419555664062, 46.35599899291992, 724.722900390625)}], 'wmode': 0, 'dir': (1.0, 0.0), 'bbox': (35.70000076293945, 712.9419555664062, 46.35599899291992, 724.722900390625)}, {'spans': [{'size': 7.0, 'flags': 4, 'bidi': 0, 'char_flags': 16, 'font': 'AndesBook', 'color': 2301728, 'alpha': 255, 'ascender': 0.9480000138282776, 'descender': -0.257999986410141, 'text': 'CIRDI RAPPORT ANNUEL 2025', 'origin': (55.0, 720.8438720703125), 'bbox': (55.0, 714.2078857421875, 152.31399536132812, 722.64990234375)}], 'wmode': 0, 'dir': (1.0, 0.0), 'bbox': (55.0, 714.2078857421875, 152.31399536132812, 722.6499

In [12]:
print(type(block))
print(block.keys())

<class 'dict'>
dict_keys(['type', 'number', 'flags', 'bbox', 'lines'])


Avec dict, le block est sous forme d'un dictionnaire. Nous avons une structure imbriquée.

PAGE -> block -> line -> span -> text

Le span va permettre d'obtenir un morceau de texte plus précis avec sa propre bbox

In [13]:
block["bbox"]

(35.70000076293945, 712.8818969726562, 1116.0, 724.722900390625)

In [14]:
lines = block["lines"]

print("Nombre de lignes: ", len(lines))
print("\n")
print("première ligne: ")
print(lines[0])

Nombre de lignes:  3


première ligne: 
{'spans': [{'size': 9.0, 'flags': 20, 'bidi': 0, 'char_flags': 24, 'font': 'AndesBold', 'color': 2301728, 'alpha': 255, 'ascender': 0.9679999947547913, 'descender': -0.3409999907016754, 'text': '22', 'origin': (35.70000076293945, 721.6539306640625), 'bbox': (35.70000076293945, 712.9419555664062, 46.35599899291992, 724.722900390625)}], 'wmode': 0, 'dir': (1.0, 0.0), 'bbox': (35.70000076293945, 712.9419555664062, 46.35599899291992, 724.722900390625)}


In [15]:
line = lines[0]

print(type(line))
print(line.keys())

<class 'dict'>
dict_keys(['spans', 'wmode', 'dir', 'bbox'])


In [16]:
# Bounding box de la ligne entière
line["bbox"]

(35.70000076293945, 712.9419555664062, 46.35599899291992, 724.722900390625)

In [17]:
# Direction du texte
line["dir"]

(1.0, 0.0)

Le texte avance horizontalement

In [18]:
# Mode d'écriture du texte, notamment horizontal/vertical
line["wmode"]

0

In [19]:
line["spans"]

[{'size': 9.0,
  'flags': 20,
  'bidi': 0,
  'char_flags': 24,
  'font': 'AndesBold',
  'color': 2301728,
  'alpha': 255,
  'ascender': 0.9679999947547913,
  'descender': -0.3409999907016754,
  'text': '22',
  'origin': (35.70000076293945, 721.6539306640625),
  'bbox': (35.70000076293945,
   712.9419555664062,
   46.35599899291992,
   724.722900390625)}]

Une ligne peut être composée de plusieurs spans parce que la mise en forme peut changer au milieu de la ligne.

In [20]:
type(line["spans"])

list

In [21]:
line["spans"][0]

{'size': 9.0,
 'flags': 20,
 'bidi': 0,
 'char_flags': 24,
 'font': 'AndesBold',
 'color': 2301728,
 'alpha': 255,
 'ascender': 0.9679999947547913,
 'descender': -0.3409999907016754,
 'text': '22',
 'origin': (35.70000076293945, 721.6539306640625),
 'bbox': (35.70000076293945,
  712.9419555664062,
  46.35599899291992,
  724.722900390625)}

In [22]:
type(line["spans"][0])

dict

Liste de dictionnaires

In [23]:
spans = line["spans"]

print("Nombre de spans :", len(spans))

span = spans[0]

print(type(span))
print(span.keys())

Nombre de spans : 1
<class 'dict'>
dict_keys(['size', 'flags', 'bidi', 'char_flags', 'font', 'color', 'alpha', 'ascender', 'descender', 'text', 'origin', 'bbox'])


In [24]:
print("Texte :", span["text"])
print("BBox :", span["bbox"])
print("Police :", span["font"])
print("Taille :", span["size"])

Texte : 22
BBox : (35.70000076293945, 712.9419555664062, 46.35599899291992, 724.722900390625)
Police : AndesBold
Taille : 9.0


span["font"] -> police

span["size"] -> taille

## Boucle qui parcourt tous les spans texte de la page

In [25]:
# Pour chaque bloc de la page
for block in page_dict["blocks"]: 
    # Si block n'est pas un texte, continue
    if block["type"] != 0:
        continue
    for line in block["lines"]:
        for span in line["spans"]:
            print(span["text"],"->",span["bbox"])

22 -> (35.70000076293945, 712.9419555664062, 46.35599899291992, 724.722900390625)
CIRDI RAPPORT ANNUEL 2025 -> (55.0, 714.2078857421875, 152.31399536132812, 722.64990234375)
23 -> (1105.14599609375, 712.8818969726562, 1116.0, 724.662841796875)
RÉPARTITION DES AFFAIRES ENREGISTRÉES AU COURS DE L’EXERCICE 2025  -> (186.052001953125, 58.70428466796875, 516.6380615234375, 71.30428314208984)
SELON LE SECTEUR ÉCONOMIQUE -> (186.052001953125, 69.70428466796875, 330.6766052246094, 82.30428314208984)
Au cours de l’exercice 2025, un total de 233 nominations ont été effectuées auprès de  -> (187.4694061279297, 448.2207946777344, 534.5057373046875, 459.0747985839844)
tribunaux, d’une commission de conciliation et de comités ad hoc CIRDI. Ces nominations  -> (187.4694061279297, 460.2178039550781, 542.1524658203125, 471.0747985839844)
ont été effectuées au sein de 51 tribunaux dans des arbitrages initiaux, au sein d’un tribunal  -> (187.47219848632812, 472.2178039550781, 542.0955200195312, 483.07180

### Potentielle extraction future

PDF -> page -> blocks -> lines -> spans -> texte + position -> regroupement intelligent -> document / chunks -> embeddings -> rag

## Bounding box autour des mots

### Transformation de la bbox en Rect

In [26]:
print(span.keys())

dict_keys(['size', 'flags', 'bidi', 'char_flags', 'font', 'color', 'alpha', 'ascender', 'descender', 'text', 'origin', 'bbox'])


In [27]:
# Transformation des nombres de la bbox en un objet représentant un rectangle géométrique

bbox = span["bbox"]

rect = pymupdf.Rect(bbox)

print(bbox)
print(rect)

(619.3390502929688, 363.1570739746094, 776.9546508789062, 373.4080810546875)
Rect(619.3390502929688, 363.1570739746094, 776.9546508789062, 373.4080810546875)


### Dessin du rectangle

In [28]:
# Création d'uen couche de dessin associée à la page
shape = page.new_shape()

# dessine le rectangle correspondant à la bbox
shape.draw_rect(rect)

# Rectangle rouge
shape.finish(
    color=(1,0,0), # RGB
    width=2
)
shape.commit() # applique le dessin à l'objet page en mémoire

### Transformation de la page en image

In [29]:
from pathlib import Path
from IPython.display import Image, display

# Dossier de sortie
PROCESSED_DIR = Path("../data/processed")

# Crée le dossier s'il n'existe pas
PROCESSED_DIR.mkdir(parents=True, exist_ok=True)

# Chemin du fichier 
output_path = PROCESSED_DIR / "page_13_bbox2.png"

# Transforme la page PDF en image
pix = page.get_pixmap(matrix=pymupdf.Matrix(1.5, 1.5))

# Enregistre l'image sur le disque
pix.save(str(output_path))

# Vérification
print("Chemin :", output_path.resolve())
print("Existe :", output_path.exists())

Chemin : C:\Users\nico_\Desktop\rag_pdf\data\processed\page_13_bbox2.png
Existe : True


In [30]:
span["text"]

'Asie du Sud & Asie de l’Est & Pacifique  11%'

### Concordance entre span["text"] et la bouding box dans l'image

In [31]:
print("SPAN")
print("texte :", span["text"])
print("bbox  :", span["bbox"])

print("\nLINE")
print("bbox :", line["bbox"])

print("\nBLOCK")
print("bbox :", block["bbox"])

SPAN
texte : Asie du Sud & Asie de l’Est & Pacifique  11%
bbox  : (619.3390502929688, 363.1570739746094, 776.9546508789062, 373.4080810546875)

LINE
bbox : (619.3390502929688, 363.1570739746094, 776.9546508789062, 373.4080810546875)

BLOCK
bbox : (619.3390502929688, 363.1570739746094, 776.9546508789062, 373.4080810546875)


## Exploration des tableaux avec PyMuPDF

In [32]:
doc = pymupdf.open(pdf_path)
print(doc)

Document('..\data\raw\worldbank_report.pdf')


In [33]:
print(len(doc))

38


### Pages contenant des tableaux

In [34]:
for page_index in range(len(doc)):
    page = doc[page_index]
    text = page.get_text("text")

    if (
        "Australie" in text
        and "Signature" in text
        and "Entrée en vigueur" in text
    ):
        print(
            "Index PyMuPDF :", page_index,
            "| page du fichier :", page_index + 1
        )

Index PyMuPDF : 34 | page du fichier : 35


In [35]:
for page_index in [34, 35, 36]:
    page = doc[page_index]

    print(f"\n{'=' * 60}")
    print(f"PAGE INDEX {page_index} | PAGE FICHIER {page_index + 1}")
    print("=" * 60)

    text = page.get_text("text")
    print(text[:3000])


PAGE INDEX 34 | PAGE FICHIER 35
66
CIRDI RAPPORT ANNUEL 2025
67
LISTE DES ÉTATS 
CONTRACTANTS ET SIGNATAIRES 
DE LA CONVENTION
AU 30 JUIN 2025
Les 165 États qui figurent sur la liste ci-dessous ont signé la Convention pour le règlement des différends relatifs aux 
investissements entre États et ressortissants d’autres États aux dates indiquées. Le nom des 158 États qui ont déposé leurs 
instruments de ratification est en caractères gras, et les dates de dépôt ainsi que d’accession au statut d’État contractant 
par l’entrée en vigueur de la Convention en ce qui concerne chacun d’eux sont également indiquées.
Afghanistan
30 sept. 1966
25 juin 1968
25 juill. 1968
Albanie
15 oct. 1991
15 oct. 1991
14 nov. 1991
Algérie
17 avr. 1995 
21 fév. 1996
22 mars 1996
Angola
14 juill. 2022
21 sep. 2022
21 oct. 2022
Argentine
21 mai 1991
19 oct. 1994
18 nov. 1994
Arménie
16 sep. 1992
16 sep. 1992
16 oct. 1992
Australie
24 mars 1975
2 mai 1991
1 juin 1991
Autriche
17 mai 1966
25 mai 1971
24 juin 1971


In [36]:
for page_index in [34, 35, 36]:
    page = doc[page_index]
    tables = page.find_tables().tables

    first_cells = []

    for table in tables:
        data = table.extract()

        if data and data[0]:
            first_cells.append(data[0][0])

    print(f"\nPage index {page_index}")
    print("Nombre de fragments :", len(tables))
    print("Premières cellules :", first_cells)

Consider using the pymupdf_layout package for a greatly improved page layout analysis.

Page index 34
Nombre de fragments : 26
Premières cellules : ['État', 'Brunéi Darussalam', 'Burkina Faso', 'Cabo Verde', 'Cameroun', 'État', 'Chili', 'Albanie', 'Chypre', 'Allemagne', 'Comores', 'Arabie saoudite', 'Congo, République du', 'Arménie', 'Costa Rica', 'Autriche', 'Croatie', 'Bahamas', 'Djibouti', 'Bangladesh', 'El Salvador', 'Bélarus', 'Équateur', 'Belize', 'Estonie', 'Bosnie-Herzégovine']

Page index 35
Nombre de fragments : 30
Premières cellules : ['État', 'État', 'Ethiopie', 'Kenya', 'Fidji', 'Kuweït', 'France', 'Lettonie', 'Gambie', 'Libéria', 'Ghana', 'Luxembourg', 'Grenade', 'Madagascar', 'Guinée', 'Malawi', 'Guinée équatoriale', 'Malte', 'Haïti', 'Maurice', 'Iles Salomon', 'Mexique', 'Irak', 'Moldavie', 'Islande', 'Monténégro', 'Italie', 'Namibie', 'Japon', 'Népal']

Page index 36
Nombre de fragments : 30
Premières cellules : ['État', 'État', 'Nigéria', 'Serbie', 'Nouvelle-Zélande',

In [37]:
page = doc[34]

page_dict = page.get_text("dict")

countries_to_check = [
    "Afghanistan",
    "Albanie",
    "Algérie",
    "Allemagne",
    "Angola",
    "Arabie saoudite",
    "Argentine",
]

for country in countries_to_check:
    found = False

    for block in page_dict["blocks"]:
        # Si block n'est pas un texte, continue
        if block["type"] != 0:
            continue

        for line in block["lines"]:
            for span in line["spans"]:

                if country in span["text"]:
                    print(
                        f"{country:20} -> "
                        f"texte={span['text']} | "
                        f"bbox={span['bbox']}"
                    )
                    found = True

    if not found:
        print(f"{country:20} -> NON TROUVÉ")

Afghanistan          -> texte=Afghanistan | bbox=(40.95759963989258, 298.1300964355469, 88.77861022949219, 309.256591796875)
Albanie              -> texte=Albanie | bbox=(40.95759963989258, 318.09320068359375, 69.85759735107422, 329.2196960449219)
Algérie              -> texte=Algérie | bbox=(41.211578369140625, 338.2467041015625, 68.27558135986328, 349.3731994628906)
Allemagne            -> texte=Allemagne | bbox=(41.20759963989258, 358.405517578125, 81.72710418701172, 369.5320129394531)
Angola               -> texte=Angola | bbox=(41.203094482421875, 378.56219482421875, 68.02909851074219, 389.6886901855469)
Arabie saoudite      -> texte=Arabie saoudite | bbox=(41.20759963989258, 398.7214050292969, 101.52361297607422, 409.847900390625)
Argentine            -> texte=Argentine | bbox=(40.948089599609375, 418.877685546875, 78.88359069824219, 430.0041809082031)


### Observation — tableau multi-page

Le tableau "Liste des États contractants et signataires de la Convention"
s'étend sur 3 pages physiques consécutives du PDF :

- index PyMuPDF 34
- index PyMuPDF 35
- index PyMuPDF 36

`page.find_tables()` ne reconstruit pas correctement le tableau complet.
Sur ce document, les lignes avec fond coloré sont principalement détectées
comme des tables indépendantes de 1 ligne × 4 colonnes, tandis que plusieurs
lignes sans fond coloré ne sont pas détectées comme tables.

En revanche, `page.get_text("dict")` récupère bien le texte des lignes
manquantes ainsi que leurs bounding boxes. Les X pour les colonnes et les y pour les lignes.

Conclusion : le contenu est disponible dans le PDF, mais une reconstruction
layout-aware est nécessaire pour obtenir le tableau logique complet.

In [38]:
page = doc[34]
page_dict = page.get_text("dict")

target_y = 298.13
tolerance = 0.5

same_row_spans = []

for block in page_dict["blocks"]:
    if block["type"] != 0:
        continue

    for line in block["lines"]:
        for span in line["spans"]:
            x0, y0, x1, y1 = span["bbox"]

            if abs(y0 - target_y) <= tolerance: # Si la distance antre les 2 valeurs inférieures à tolérance
                same_row_spans.append({ # Appartiennent à la même ligne
                    "text": span["text"],
                    "x0": x0,
                    "y0": y0,
                    "bbox": span["bbox"],
                })

same_row_spans = sorted(
    same_row_spans,
    key=lambda span: span["x0"]  # on trie de gauche à droite selon leur x0, l'ordre spatial
)

for span in same_row_spans:
    print(
        f"x={span['x0']:.1f} | "
        f"y={span['y0']:.1f} | "
        f"text={span['text']!r}"
    )

x=41.0 | y=298.1 | text='Afghanistan'
x=242.2 | y=298.3 | text='30 sept. 1966'
x=342.9 | y=298.3 | text='25 juin 1968'
x=443.5 | y=298.3 | text='25 juill. 1968'


In [39]:
page = doc[34]
page_dict = page.get_text("dict")

left_spans = []

for block in page_dict["blocks"]:
    if block["type"] != 0:
        continue

    for line in block["lines"]:
        for span in line["spans"]:
            x0, y0, x1, y1 = span["bbox"]

            if x0 < 550:
                left_spans.append({
                    "text": span["text"].strip(), # etire les espaces inutiles au début et à la fin du texte
                    "x0": x0,
                    "y0": y0,
                    "x1": x1,
                    "y1": y1,
                    "bbox": span["bbox"],
                })

In [40]:
table_spans = [
    span
    for span in left_spans
    if 290 <= span["y0"] <= 700 # Garde les spans du panneau gauche dont la position verticale se situe entre 290 et 700
]

print("Nombre de spans :", len(table_spans))

Nombre de spans : 78


In [41]:
rows = []
tolerance = 0.5

for span in sorted(table_spans, key=lambda s: s["y0"]): # on trie tous les spans du haut vers le bas

    matching_row = None

    for row in rows:
        if abs(span["y0"] - row["y"]) <= tolerance: # y a t il une ligne ayant approximativement le même y?
            matching_row = row
            break # si oui, stoppe la recherche

    if matching_row is None: # Si aucune ligne ne correspond
        rows.append({        # On crée une nouvelle ligne
            "y": span["y0"],
            "spans": [span],
        })
    else:
        matching_row["spans"].append(span) # on ajoute le span à la ligne existante

### Concept:

Afghanistan y=298.1 -> aucune row proche -> création ROW y≈298 -> 30 sept. y=298.3 -> ROW 298 existe / différence = 0.2 -> ajout dans ROW 298 -> 25 juin y=298.3 -> ajout dans ROW 298 -> Albanie y=318.1 -> aucune ROW proche -> création nouvelle ROW

In [42]:
for row in rows:
    row["spans"] = sorted(
        row["spans"],
        key=lambda s: s["x0"]
    )

    texts = [
        span["text"]
        for span in row["spans"]
        if span["text"]
    ]

    print(f"y={row['y']:.1f} -> {texts}")

y=298.1 -> ['Afghanistan', '30 sept. 1966', '25 juin 1968', '25 juill. 1968']
y=318.1 -> ['Albanie', '15 oct. 1991', '15 oct. 1991', '14 nov. 1991']
y=338.2 -> ['Algérie', '17 avr. 1995', '21 fév. 1996', '22 mars 1996']
y=358.4 -> ['Allemagne', '27 janv. 1966', '18 avr. 1969', '18 mai 1969']
y=378.6 -> ['Angola', '14 juill. 2022', '21 sep. 2022', '21 oct. 2022']
y=398.7 -> ['Arabie saoudite', '28 sep. 1979', '8 mai 1980', '7 juin 1980']
y=418.9 -> ['Argentine', '21 mai 1991', '19 oct. 1994', '18 nov. 1994']
y=439.0 -> ['Arménie', '16 sep. 1992', '16 sep. 1992', '16 oct. 1992']
y=459.2 -> ['Australie', '24 mars 1975', '2 mai 1991', '1 juin 1991']
y=479.3 -> ['Autriche', '17 mai 1966', '25 mai 1971', '24 juin 1971']
y=499.5 -> ['Azerbaïdjan', '18 sep. 1992', '18 sep. 1992', '18 oct. 1992']
y=519.7 -> ['Bahamas', '19 oct. 1995', '19 oct. 1995', '18 nov. 1995']
y=539.8 -> ['Bahreïn', '22 sep. 1995', '14 fév. 1996', '15 mars 1996']
y=560.0 -> ['Bangladesh', '20 nov. 1979', '27 mars 1980', '

## Colonnes

In [43]:
column_ranges = [
    (0, 200),      # colonne 0 : État
    (200, 300),    # colonne 1 : Signature
    (300, 400),    # colonne 2 : Ratification
    (400, 550),    # colonne 3 : Entrée en vigueur
]

In [44]:
def get_column_index(x0):
    for column_index, (x_min, x_max) in enumerate(column_ranges):
        if x_min <= x0 < x_max:
            return column_index

    return None

### Construire quatre cellules pour chaque ligne

In [45]:
structured_rows = []

for row in rows:

    cells = [None, None, None, None]

    for span in row["spans"]:
        column_index = get_column_index(span["x0"])

        if column_index is not None:
            cells[column_index] = span["text"]

    structured_rows.append(cells)

In [46]:
for row in structured_rows:
    print(row)

['Afghanistan', '30 sept. 1966', '25 juin 1968', '25 juill. 1968']
['Albanie', '15 oct. 1991', '15 oct. 1991', '14 nov. 1991']
['Algérie', '17 avr. 1995', '21 fév. 1996', '22 mars 1996']
['Allemagne', '27 janv. 1966', '18 avr. 1969', '18 mai 1969']
['Angola', '14 juill. 2022', '21 sep. 2022', '21 oct. 2022']
['Arabie saoudite', '28 sep. 1979', '8 mai 1980', '7 juin 1980']
['Argentine', '21 mai 1991', '19 oct. 1994', '18 nov. 1994']
['Arménie', '16 sep. 1992', '16 sep. 1992', '16 oct. 1992']
['Australie', '24 mars 1975', '2 mai 1991', '1 juin 1991']
['Autriche', '17 mai 1966', '25 mai 1971', '24 juin 1971']
['Azerbaïdjan', '18 sep. 1992', '18 sep. 1992', '18 oct. 1992']
['Bahamas', '19 oct. 1995', '19 oct. 1995', '18 nov. 1995']
['Bahreïn', '22 sep. 1995', '14 fév. 1996', '15 mars 1996']
['Bangladesh', '20 nov. 1979', '27 mars 1980', '26 avr. 1980']
['Barbade', '13 mai 1981', '1 nov. 1983', '1 déc. 1983']
['Bélarus', '10 juill. 1992', '10 juill. 1992', '9 août 1992']
['Belgique', '15 dé

In [47]:
page = doc[36]
page_dict = page.get_text("dict")

search_terms = [
    "Royaume-Uni",
    "Grande-Bretagne",
    "Irlande du Nord",
]

for block in page_dict["blocks"]:
    if block["type"] != 0:
        continue

    for line in block["lines"]:
        for span in line["spans"]:
            text = span["text"].strip()

            if any(term in text for term in search_terms):
                print(
                    f"text={text!r}\n"
                    f"bbox={span['bbox']}\n"
                    f"line_bbox={line['bbox']}\n"
                    f"block_bbox={block['bbox']}\n"
                    f"{'-' * 70}"
                )

text='Royaume-Uni de Grande-Bretagne et'
bbox=(40.5, 515.915771484375, 183.92898559570312, 527.0422973632812)
line_bbox=(40.5, 515.915771484375, 183.92898559570312, 527.0422973632812)
block_bbox=(40.5, 515.915771484375, 491.6272277832031, 536.0437622070312)
----------------------------------------------------------------------
text='Irlande du Nord'
bbox=(47.1693000793457, 524.915771484375, 105.97230529785156, 536.0422973632812)
line_bbox=(40.5, 522.3756713867188, 105.97230529785156, 536.0437622070312)
block_bbox=(40.5, 515.915771484375, 491.6272277832031, 536.0437622070312)
----------------------------------------------------------------------


PyMuPDF a séparé le nom en deux spans situés sur deux lignes verticales différentes

In [48]:
page = doc[36]
page_dict = page.get_text("dict")

for block in page_dict["blocks"]:
    block_text = []

    if block["type"] != 0:
        continue

    for line in block["lines"]:
        for span in line["spans"]:
            block_text.append(span["text"].strip())

    if any("Royaume-Uni" in text for text in block_text):
        print("BLOCK BBOX:", block["bbox"])

        for line in block["lines"]:
            print("\nLINE BBOX:", line["bbox"])

            for span in line["spans"]:
                print(
                    f"  text={span['text']!r} "
                    f"bbox={span['bbox']}"
                )

BLOCK BBOX: (40.5, 515.915771484375, 491.6272277832031, 536.0437622070312)

LINE BBOX: (40.5, 515.915771484375, 183.92898559570312, 527.0422973632812)
  text='Royaume-Uni de Grande-Bretagne et ' bbox=(40.5, 515.915771484375, 183.92898559570312, 527.0422973632812)

LINE BBOX: (40.5, 522.3756713867188, 105.97230529785156, 536.0437622070312)
  text='d' bbox=(40.5, 524.917236328125, 45.33650207519531, 536.0437622070312)
  text=',' bbox=(45.33639907836914, 522.3756713867188, 47.169349670410156, 531.533447265625)
  text='Irlande du Nord' bbox=(47.1693000793457, 524.915771484375, 105.97230529785156, 536.0422973632812)

LINE BBOX: (241.7667999267578, 520.5432739257812, 288.6103210449219, 530.998291015625)
  text='26 mai 1965' bbox=(241.7667999267578, 520.5432739257812, 288.6103210449219, 530.998291015625)

LINE BBOX: (342.40020751953125, 520.5432739257812, 389.6517028808594, 530.998291015625)
  text='19 déc. 1966' bbox=(342.40020751953125, 520.5432739257812, 389.6517028808594, 530.998291015625

In [49]:
page = doc[34]
page_dict = page.get_text("dict")

for block in page_dict["blocks"]:
    block_text = []

    if block["type"] != 0:
        continue

    for line in block["lines"]:
        for span in line["spans"]:
            block_text.append(span["text"].strip())

    if any("Afghanistan" in text for text in block_text):
        print("BLOCK BBOX:", block["bbox"])

        for line in block["lines"]:
            print("\nLINE BBOX:", line["bbox"])

            for span in line["spans"]:
                print(
                    f"  text={span['text']!r} "
                    f"bbox={span['bbox']}"
                )

BLOCK BBOX: (40.95759963989258, 298.1300964355469, 491.42364501953125, 309.256591796875)

LINE BBOX: (40.95759963989258, 298.1300964355469, 88.77861022949219, 309.256591796875)
  text='Afghanistan' bbox=(40.95759963989258, 298.1300964355469, 88.77861022949219, 309.256591796875)

LINE BBOX: (242.22059631347656, 298.2575988769531, 294.3757629394531, 308.71258544921875)
  text='30 sept. 1966' bbox=(242.22059631347656, 298.2575988769531, 294.3757629394531, 308.71258544921875)

LINE BBOX: (342.85211181640625, 298.2575988769531, 389.39813232421875, 308.71258544921875)
  text='25 juin 1968' bbox=(342.85211181640625, 298.2575988769531, 389.39813232421875, 308.71258544921875)

LINE BBOX: (443.4836120605469, 298.2575988769531, 491.42364501953125, 308.71258544921875)
  text='25 juill. 1968' bbox=(443.4836120605469, 298.2575988769531, 491.42364501953125, 308.71258544921875)


## Code après expériences

In [ ]:
# page index 34, 35ème page
page = doc[34]

# Représentation structurée de la page
page_dict = page.get_text("dict")

table_blocks = [] # Liste qui comprendra les blocks appartenant au tableau

# Pour chaque block
for block in page_dict["blocks"]:
    # Accepte que le texte, sinon continue
    if block["type"] != 0:
        continue

    # Coordonnées bbox du block entier
    x0, y0, x1, y1 = block["bbox"]

    # Zone expérimentale :
    # panneau gauche + zone des lignes du tableau
    if x0 < 550 and 290 <= y0 <= 700:

        spans_data = [] # Liste comprenant text et ses coordonnées bbox

        # A l'intérieur du block, descente dans la hiérarchie jusqu'aux spam
        for line in block["lines"]:
            for span in line["spans"]:
                text = span["text"].strip() # enlève les espaces inutiles au début et à la fin

                if text: # évite de stocker des spans vides
                    spans_data.append({
                        "text": text,
                        "bbox": span["bbox"]
                    })

        table_blocks.append({
            "bbox": block["bbox"],
            "spans": spans_data,
        })

# Triage des blocks du plus petit au plus grand en fonction de y0
table_blocks = sorted(
    table_blocks,
    key=lambda block: block["bbox"][1] # [1] correspond à y0. 
)

# La bbox du block peut aider à comprendre la structure du tableau, tandis que les bbox des spans pourront servir à faire un surlignage beaucoup plus précis dans PDF.js

for block in table_blocks:
    print(block)

{'bbox': (40.95759963989258, 298.1300964355469, 491.42364501953125, 309.256591796875), 'spans': [{'text': 'Afghanistan', 'bbox': (40.95759963989258, 298.1300964355469, 88.77861022949219, 309.256591796875)}, {'text': '30 sept. 1966', 'bbox': (242.22059631347656, 298.2575988769531, 294.3757629394531, 308.71258544921875)}, {'text': '25 juin 1968', 'bbox': (342.85211181640625, 298.2575988769531, 389.39813232421875, 308.71258544921875)}, {'text': '25 juill. 1968', 'bbox': (443.4836120605469, 298.2575988769531, 491.42364501953125, 308.71258544921875)}]}
{'bbox': (40.95759963989258, 318.09320068359375, 488.8291931152344, 329.2196960449219), 'spans': [{'text': 'Albanie', 'bbox': (40.95759963989258, 318.09320068359375, 69.85759735107422, 329.2196960449219)}, {'text': '15 oct. 1991', 'bbox': (242.22430419921875, 318.220703125, 286.7983093261719, 328.6756896972656)}, {'text': '15 oct. 1991', 'bbox': (342.8576965332031, 318.220703125, 387.43170166015625, 328.6756896972656)}, {'text': '14 nov. 1991

In [64]:
print(table_blocks[0])

{'bbox': (40.95759963989258, 298.1300964355469, 491.42364501953125, 309.256591796875), 'spans': [{'text': 'Afghanistan', 'bbox': (40.95759963989258, 298.1300964355469, 88.77861022949219, 309.256591796875)}, {'text': '30 sept. 1966', 'bbox': (242.22059631347656, 298.2575988769531, 294.3757629394531, 308.71258544921875)}, {'text': '25 juin 1968', 'bbox': (342.85211181640625, 298.2575988769531, 389.39813232421875, 308.71258544921875)}, {'text': '25 juill. 1968', 'bbox': (443.4836120605469, 298.2575988769531, 491.42364501953125, 308.71258544921875)}]}


In [52]:
table_blocks

[{'bbox': (40.95759963989258,
   298.1300964355469,
   491.42364501953125,
   309.256591796875),
  'spans': [{'text': 'Afghanistan',
    'bbox': (40.95759963989258,
     298.1300964355469,
     88.77861022949219,
     309.256591796875)},
   {'text': '30 sept. 1966',
    'bbox': (242.22059631347656,
     298.2575988769531,
     294.3757629394531,
     308.71258544921875)},
   {'text': '25 juin 1968',
    'bbox': (342.85211181640625,
     298.2575988769531,
     389.39813232421875,
     308.71258544921875)},
   {'text': '25 juill. 1968',
    'bbox': (443.4836120605469,
     298.2575988769531,
     491.42364501953125,
     308.71258544921875)}]},
 {'bbox': (40.95759963989258,
   318.09320068359375,
   488.8291931152344,
   329.2196960449219),
  'spans': [{'text': 'Albanie',
    'bbox': (40.95759963989258,
     318.09320068359375,
     69.85759735107422,
     329.2196960449219)},
   {'text': '15 oct. 1991',
    'bbox': (242.22430419921875,
     318.220703125,
     286.7983093261719,
     3

In [56]:
first_block = table_blocks[0]

print("BBOX du block :", first_block["bbox"])

for span in first_block["spans"]:
    x0,y0,x1,y1 = span["bbox"]

    print(span["text"],
          "->",
          f"x0={x0:.1f},yo={y0:.1f},x1={x1:.1f},y1={y1:.1f}")

# x0 -> gauche | y0 -> haut | x1 -> droite | y1 -> bas

BBOX du block : (40.95759963989258, 298.1300964355469, 491.42364501953125, 309.256591796875)
Afghanistan -> x0=41.0,yo=298.1,x1=88.8,y1=309.3
30 sept. 1966 -> x0=242.2,yo=298.3,x1=294.4,y1=308.7
25 juin 1968 -> x0=342.9,yo=298.3,x1=389.4,y1=308.7
25 juill. 1968 -> x0=443.5,yo=298.3,x1=491.4,y1=308.7


In [60]:
for block in table_blocks[:5]:
    print("---")

    for span in block["spans"]:
        x0 = span["bbox"][0]

        print(
            f"x0={x0:.1f} -> {span['text']}"
        )

---
x0=41.0 -> Afghanistan
x0=242.2 -> 30 sept. 1966
x0=342.9 -> 25 juin 1968
x0=443.5 -> 25 juill. 1968
---
x0=41.0 -> Albanie
x0=242.2 -> 15 oct. 1991
x0=342.9 -> 15 oct. 1991
x0=443.5 -> 14 nov. 1991
---
x0=41.2 -> Algérie
x0=242.5 -> 17 avr. 1995
x0=343.1 -> 21 fév. 1996
x0=443.7 -> 22 mars 1996
---
x0=41.2 -> Allemagne
x0=242.5 -> 27 janv. 1966
x0=343.1 -> 18 avr. 1969
x0=443.7 -> 18 mai 1969
---
x0=41.2 -> Angola
x0=242.5 -> 14 juill. 2022
x0=343.1 -> 21 sep. 2022
x0=443.7 -> 21 oct. 2022


In [66]:
# Rechrche de Belize avec valeurs manquantes

for block in table_blocks:
    for span in block["spans"]:
        if span["text"] == "Belize":
            print(block)

{'bbox': (40.95759963989258, 640.7037963867188, 288.9147644042969, 651.1588134765625), 'spans': [{'text': 'Belize', 'bbox': (40.95759963989258, 640.7037963867188, 62.8026008605957, 651.1588134765625)}, {'text': '19 déc. 1986', 'bbox': (242.22430419921875, 640.746337890625, 288.9147644042969, 650.997314453125)}]}


## Algorithme

### Colonnes

### Position des x0

In [73]:
x_position = []

for block in table_blocks:
    for span in block["spans"]:
        x0 = span["bbox"][0]
        x_position.append(x0)

In [68]:
x_position

[40.95759963989258,
 242.22059631347656,
 342.85211181640625,
 443.4836120605469,
 40.95759963989258,
 242.22430419921875,
 342.8576965332031,
 443.4910888671875,
 41.211578369140625,
 242.47457885742188,
 343.1060791015625,
 443.7375793457031,
 41.20759963989258,
 242.47430419921875,
 343.1076965332031,
 443.7410888671875,
 41.203094482421875,
 242.46609497070312,
 343.09759521484375,
 443.7290954589844,
 41.20759963989258,
 242.47430419921875,
 343.1076965332031,
 443.7410888671875,
 40.948089599609375,
 242.21109008789062,
 342.84259033203125,
 443.4740905761719,
 40.95759963989258,
 242.22430419921875,
 342.8576965332031,
 443.4910888671875,
 40.95660400390625,
 242.2196044921875,
 342.8511047363281,
 443.48260498046875,
 40.95759963989258,
 242.22430419921875,
 342.8576965332031,
 443.4910888671875,
 40.95660400390625,
 242.2196044921875,
 342.8511047363281,
 443.48260498046875,
 40.95759963989258,
 242.22430419921875,
 342.8576965332031,
 443.4910888671875,
 40.95660400390625,
 2

### Tri du plus petit au plus grand

In [72]:
x_positions_sorted = sorted(x_position)
print(x_positions_sorted[:20])

[40.948089599609375, 40.95660400390625, 40.95660400390625, 40.95660400390625, 40.95660400390625, 40.95660400390625, 40.95759963989258, 40.95759963989258, 40.95759963989258, 40.95759963989258, 40.95759963989258, 40.95759963989258, 40.95759963989258, 40.95759963989258, 40.95759963989258, 40.95759963989258, 41.203094482421875, 41.20759963989258, 41.20759963989258, 41.211578369140625]


# Mesure des écarts entre les x

In [74]:
x_gaps = []

for i in range(len(x_positions_sorted) -1): # -1 car dans la boucle on utilise x_positions_sorted[i + 1] et si on allait jusqu’au dernier élément, il n’existerait plus d’élément suivant
    gap = x_positions_sorted[i+1] - x_positions_sorted[i]
    x_gaps.append(gap)

print(x_gaps[:20])


[0.008514404296875, 0.0, 0.0, 0.0, 0.0, 0.000995635986328125, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.24549484252929688, 0.004505157470703125, 0.0, 0.003978729248046875, 200.99951171875]


Beaucoup de positions sont extrêmement proches puis pour le dernier élément, un écart de 201 points

In [ ]:
# Tri du plus grand au plus petit
largest_gaps = sorted(x_gaps, reverse=True)

print(largest_gaps[:10])

[200.99951171875, 100.36801147460938, 100.36639404296875, 0.24549484252929688, 0.241790771484375, 0.239898681640625, 0.238006591796875, 0.008514404296875, 0.008514404296875, 0.008514404296875]


Il y a une rupture entre 100.36 et 0.24.
Ls coordionnées commencenr à révéler la structure

### Affichage des gros écarts avec leurs positions

In [77]:
for i in range(len(x_positions_sorted) -1):
    x1 = x_positions_sorted[i]
    x2 = x_positions_sorted[i+1]
    gap = x2 - x1

    if gap > 1:
        print (f"{x1:.1f} -> {x2:.1f} : gap = {gap:.1f}")

41.2 -> 242.2 : gap = 201.0
242.5 -> 342.8 : gap = 100.4
343.1 -> 443.5 : gap = 100.4


### Fonction avec tolérance

In [ ]:
def group_close_positions(positions, tolerance):
    """
    positions : liste de x triés
    tolerance : distance maximale qu'on accepte entre deux positions voisines pour considérer qu'elles appartiennt au même groupe
    """
    groups = [[positions[0]]] # premier élément de la liste et on veut une liste de groupe

    for x in positions[1:]: # commence au 2ème élément de la liste
        if x - groups[-1][-1] <= tolerance: # prend le dernier groupe et le dernier élément du dernier groupe
            groups[-1].append(x) # ajoute x au groupe actuel

        else:
            groups.append([x]) # sinon crée un nouveau groupe

    return groups

In [90]:
groups = group_close_positions(x_positions_sorted, tolerance=1)

for group in groups:
    print(len(group), round(sum(group) / len(group), 1))

20 41.0
20 242.3
19 342.9
19 443.5


La fonction a découvert 4 groupes de positions, donc les 4 colonnes du panneau.

Pour les deux derniers, la longueur est de 19 car pour Bélize, les deux dernières colonnes sont vides.

In [86]:
# Même fonction sans trier avant

def group_close_positions(positions, tolerance):
    positions = sorted(positions)
    groups = [[positions[0]]]

    for x in positions[1:]:
        if x - groups[-1][-1] <= tolerance:
            groups[-1].append(x)
        else:
            groups.append([x])

    return groups

In [88]:
group_close_positions(x_position, tolerance=1)

[[40.948089599609375,
  40.95660400390625,
  40.95660400390625,
  40.95660400390625,
  40.95660400390625,
  40.95660400390625,
  40.95759963989258,
  40.95759963989258,
  40.95759963989258,
  40.95759963989258,
  40.95759963989258,
  40.95759963989258,
  40.95759963989258,
  40.95759963989258,
  40.95759963989258,
  40.95759963989258,
  41.203094482421875,
  41.20759963989258,
  41.20759963989258,
  41.211578369140625],
 [242.21109008789062,
  242.2196044921875,
  242.2196044921875,
  242.2196044921875,
  242.2196044921875,
  242.2196044921875,
  242.22059631347656,
  242.22059631347656,
  242.22430419921875,
  242.22430419921875,
  242.22430419921875,
  242.22430419921875,
  242.22430419921875,
  242.22430419921875,
  242.22430419921875,
  242.22430419921875,
  242.46609497070312,
  242.47430419921875,
  242.47430419921875,
  242.47457885742188],
 [342.84259033203125,
  342.8511047363281,
  342.8511047363281,
  342.8511047363281,
  342.8511047363281,
  342.8511047363281,
  342.8521118

### Définition de la tolérance

In [ ]:
# On retire les gaps nuls pour éviter la division par 0

non_zero_gaps = [ gap for gap in largest_gaps if gap > 0]

In [94]:
# Comparaison de chaque grand gap au suivant

gap_ratios = [] # liste qui va contenir des rapports, et non plus des distances

for i in range(len(non_zero_gaps) - 1):
    ratio = non_zero_gaps[i] / non_zero_gaps[i + 1]
    gap_ratios.append(ratio)

print(gap_ratios[:10])

[2.002625226560336, 1.0000161152710132, 408.83300442856034, 1.0153193234885776, 1.0078870372726116, 1.007949737145788, 27.953405017921146, 1.0, 1.0, 1.0]


Deux ruptures, 408 et 27

In [95]:
# Plus grand ratio et sa position

max_ratio = max(gap_ratios)
rupture_index = gap_ratios.index(max_ratio)

print("Plus grans ration :", max_ratio)
print("Index de rupture :", rupture_index)

Plus grans ration : 408.83300442856034
Index de rupture : 2


In [97]:
large_gap = non_zero_gaps[rupture_index]
small_gap = non_zero_gaps[rupture_index + 1]

auto_tolerance = (large_gap + small_gap) / 2

print("Large gap :", large_gap)
print("Small gap :", small_gap)
print("Tolérance automatique :", auto_tolerance)

Large gap : 100.36639404296875
Small gap : 0.24549484252929688
Tolérance automatique : 50.30594444274902


Deux positions x séparées de moins de 50 points seont considérés comme appartenant au même groupe, au delà on crée un nouveau groupe

In [99]:
def group_close_positions(positions, tolerance):
    positions = sorted(positions)
    groups = [[positions[0]]]

    for x in positions[1:]:
        if x - groups[-1][-1] <= tolerance:
            groups[-1].append(x)
        else:
            groups.append([x])

    return groups

group_close_positions(x_position, tolerance=auto_tolerance)

[[40.948089599609375,
  40.95660400390625,
  40.95660400390625,
  40.95660400390625,
  40.95660400390625,
  40.95660400390625,
  40.95759963989258,
  40.95759963989258,
  40.95759963989258,
  40.95759963989258,
  40.95759963989258,
  40.95759963989258,
  40.95759963989258,
  40.95759963989258,
  40.95759963989258,
  40.95759963989258,
  41.203094482421875,
  41.20759963989258,
  41.20759963989258,
  41.211578369140625],
 [242.21109008789062,
  242.2196044921875,
  242.2196044921875,
  242.2196044921875,
  242.2196044921875,
  242.2196044921875,
  242.22059631347656,
  242.22059631347656,
  242.22430419921875,
  242.22430419921875,
  242.22430419921875,
  242.22430419921875,
  242.22430419921875,
  242.22430419921875,
  242.22430419921875,
  242.22430419921875,
  242.46609497070312,
  242.47430419921875,
  242.47430419921875,
  242.47457885742188],
 [342.84259033203125,
  342.8511047363281,
  342.8511047363281,
  342.8511047363281,
  342.8511047363281,
  342.8511047363281,
  342.8521118

In [100]:
for group in groups:
    print(len(group), round(sum(group) / len(group), 1))

20 41.0
20 242.3
19 342.9
19 443.5


Avec le ratio, on retrouve bien les 4 colonnes

### Transformation des groupes en une position représentative par colonne

In [ ]:
column_positions = [] # liste dune position représentative pour chaque colonne

for group in groups:
    mean_x = sum(group) / len(group)
    column_positions.append(mean_x)

print(column_positions)

[41.006848907470705, 242.27170104980468, 342.906593724301, 443.53899022152547]


Moyennes des positions de référence de chaque groupe (colonnes)

In [116]:
table_blocks

[{'bbox': (40.95759963989258,
   298.1300964355469,
   491.42364501953125,
   309.256591796875),
  'spans': [{'text': 'Afghanistan',
    'bbox': (40.95759963989258,
     298.1300964355469,
     88.77861022949219,
     309.256591796875)},
   {'text': '30 sept. 1966',
    'bbox': (242.22059631347656,
     298.2575988769531,
     294.3757629394531,
     308.71258544921875)},
   {'text': '25 juin 1968',
    'bbox': (342.85211181640625,
     298.2575988769531,
     389.39813232421875,
     308.71258544921875)},
   {'text': '25 juill. 1968',
    'bbox': (443.4836120605469,
     298.2575988769531,
     491.42364501953125,
     308.71258544921875)}]},
 {'bbox': (40.95759963989258,
   318.09320068359375,
   488.8291931152344,
   329.2196960449219),
  'spans': [{'text': 'Albanie',
    'bbox': (40.95759963989258,
     318.09320068359375,
     69.85759735107422,
     329.2196960449219)},
   {'text': '15 oct. 1991',
    'bbox': (242.22430419921875,
     318.220703125,
     286.7983093261719,
     3

In [117]:
table_blocks[-1].keys()

dict_keys(['bbox', 'spans'])

In [118]:
table_blocks[-1]["spans"]

[{'text': 'Bosnie-Herzégovine',
  'bbox': (40.95759963989258,
   680.8846435546875,
   116.87109375,
   692.0111694335938)},
 {'text': '25 avr. 1997',
  'bbox': (242.22430419921875,
   681.0121459960938,
   288.4557800292969,
   691.4671630859375)},
 {'text': '14 mai 1997',
  'bbox': (342.8576965332031,
   681.0121459960938,
   387.55072021484375,
   691.4671630859375)},
 {'text': '13 juin 1997',
  'bbox': (443.4910888671875,
   681.0121459960938,
   487.9800720214844,
   691.4671630859375)}]

In [113]:
# Recherche avec Bélize avec valeurs manquantes
# table_blocks -> bbox(block) + span/text/bbox 

belize_block = None

for block in table_blocks:
    texts = [span["text"] for span in block["spans"]]

    if "Belize" in texts:
        belize_block = block
        break

In [114]:
print(belize_block)

{'bbox': (40.95759963989258, 640.7037963867188, 288.9147644042969, 651.1588134765625), 'spans': [{'text': 'Belize', 'bbox': (40.95759963989258, 640.7037963867188, 62.8026008605957, 651.1588134765625)}, {'text': '19 déc. 1986', 'bbox': (242.22430419921875, 640.746337890625, 288.9147644042969, 650.997314453125)}]}


In [120]:
belize_block["spans"]

[{'text': 'Belize',
  'bbox': (40.95759963989258,
   640.7037963867188,
   62.8026008605957,
   651.1588134765625)},
 {'text': '19 déc. 1986',
  'bbox': (242.22430419921875,
   640.746337890625,
   288.9147644042969,
   650.997314453125)}]

In [ ]:
for span in belize_block["spans"]:
    x0 = span["bbox"][0]

    # Quelles colonnes détectées (column_positions) est la plus proche de x0
    closest_column = min(
        column_positions,
        key=lambda column_x: abs(x0 - column_x) # column_x -> chaque valeurs de column_positions sont envoyé à lambda. min() prend la valeur la plus petite
    )

    print(span["text"], "->", closest_column)

Belize -> 41.006848907470705
19 déc. 1986 -> 242.27170104980468


In [131]:
# Création de cellules vides

row = [None] * len(column_positions)

print(row)

[None, None, None, None]


41.0068 -> colonne d'index 0

242.27  -> colonne d'index 1

342.91  -> colonne d'index 2

443.54  -> colonne d'index 3

In [ ]:
for span in belize_block["spans"]:
    x0 = span["bbox"][0]

    column_index = min(
        range(len(column_positions)), # 0,1,2,3
        key=lambda i: abs(x0 - column_positions[i])
    )

    print(span["text"], "-> colonne", column_index)

Belize -> colonne 0
19 déc. 1986 -> colonne 1


In [133]:
for span in belize_block["spans"]:
    x0 = span["bbox"][0]

    column_index = min(
        range(len(column_positions)), # 0,1,2,3
        key=lambda i: abs(x0 - column_positions[i])
    )

    row[column_index] = span["text"]

In [134]:
print(row)

['Belize', '19 déc. 1986', None, None]


les deux valeurs manquantes dans le tableau au niveau de Belize sont maintenant visibles dans la reconstruction

### Même logique aux 20 lignes de table_rocks

In [ ]:
rows = [] # Lignes du tableau complet

for block in table_blocks:
    row = [None] * len(column_positions)

    for span in block["spans"]:
        x0 = span["bbox"][0]

        column_index = min(
                range(len(column_positions)), # 0,1,2,3
                key=lambda i: abs(x0 - column_positions[i])
            )

        row[column_index] = span["text"] # Remplacement par index

    rows.append(row)

    

In [136]:
for row in rows:
    print(row)

['Afghanistan', '30 sept. 1966', '25 juin 1968', '25 juill. 1968']
['Albanie', '15 oct. 1991', '15 oct. 1991', '14 nov. 1991']
['Algérie', '17 avr. 1995', '21 fév. 1996', '22 mars 1996']
['Allemagne', '27 janv. 1966', '18 avr. 1969', '18 mai 1969']
['Angola', '14 juill. 2022', '21 sep. 2022', '21 oct. 2022']
['Arabie saoudite', '28 sep. 1979', '8 mai 1980', '7 juin 1980']
['Argentine', '21 mai 1991', '19 oct. 1994', '18 nov. 1994']
['Arménie', '16 sep. 1992', '16 sep. 1992', '16 oct. 1992']
['Australie', '24 mars 1975', '2 mai 1991', '1 juin 1991']
['Autriche', '17 mai 1966', '25 mai 1971', '24 juin 1971']
['Azerbaïdjan', '18 sep. 1992', '18 sep. 1992', '18 oct. 1992']
['Bahamas', '19 oct. 1995', '19 oct. 1995', '18 nov. 1995']
['Bahreïn', '22 sep. 1995', '14 fév. 1996', '15 mars 1996']
['Bangladesh', '20 nov. 1979', '27 mars 1980', '26 avr. 1980']
['Barbade', '13 mai 1981', '1 nov. 1983', '1 déc. 1983']
['Bélarus', '10 juill. 1992', '10 juill. 1992', '9 août 1992']
['Belgique', '15 dé

## Reconnaissance d'un tableau dans une page avec des paragraphes avant ou après

In [154]:
page = doc[34]
page_dict = page.get_text("dict")

block_x_positions = []

for block in page_dict["blocks"]:
    if block["type"] != 0:
        continue

    x_positions = []

    for line in block["lines"]:
        for span in line["spans"]:
            text = span["text"].strip()

            if text:
                x_positions.append(span["bbox"][0])

    if x_positions:
        block_x_positions.append(x_positions)

for positions in block_x_positions:
    print([round(x, 1) for x in positions])

[35.7, 55.0, 1105.5]
[36.4, 36.4, 36.4]
[37.3]
[36.3, 36.3, 36.3, 36.3]
[41.0, 242.2, 342.9, 443.5]
[41.0, 242.2, 342.9, 443.5]
[41.2, 242.5, 343.1, 443.7]
[41.2, 242.5, 343.1, 443.7]
[40.9, 242.2, 342.8, 443.5]
[41.0, 242.2, 342.9, 443.5]
[41.0, 242.2, 342.9, 443.5]
[41.0, 242.2, 342.9, 443.5]
[41.0, 242.2, 342.9, 443.5]
[41.0, 242.2, 342.9, 443.5]
[41.0, 242.2, 342.9, 443.5]
[41.0, 242.2, 342.9, 443.5]
[41.0, 242.2, 342.9, 443.5]
[41.0, 242.2, 342.9, 443.5]
[41.0, 242.2, 342.9, 443.5]
[41.0, 242.2]
[41.0, 242.2, 342.9, 443.5]
[41.0, 242.2, 342.9, 443.5]
[41.2, 242.5, 343.1, 443.7]
[41.2, 242.5, 343.1, 443.7]
[41.5, 242.7]
[343.4, 343.4, 343.4]
[444.0, 444.0]
[617.1, 818.3, 919.0, 1019.6]
[617.1, 818.3, 919.0, 1019.6]
[617.1, 818.3, 919.0, 1019.6]
[617.1, 818.3, 919.0, 1019.6]
[617.1, 818.3, 919.0, 1019.6]
[617.1, 818.3, 919.0, 1019.6]
[617.1, 818.3, 919.0, 1019.6]
[617.1, 818.3, 919.0, 1019.6]
[617.1, 818.3, 919.0, 1019.6]
[617.1, 818.3, 919.0, 1019.6]
[617.1, 818.3, 919.0, 1019.6]
[

In [156]:
first_row = block_x_positions[4]
second_row = block_x_positions[5]

print(first_row)
print(second_row)

[40.95759963989258, 242.22059631347656, 342.85211181640625, 443.4836120605469]
[40.95759963989258, 242.22430419921875, 342.8576965332031, 443.4910888671875]


In [ ]:
for x1, x2 in zip(first_row, second_row): # zip() pour prendre les paires
    difference = abs(x1 - x2)
    print(difference)

0.0
0.0037078857421875
0.005584716796875
0.007476806640625


### Comparaison automatique entre un block et le block suivant

In [179]:
def compare_x_positions(position1, positions2):
    if len(position1) != len(positions2): # si les deux blocks n'ont pas le même nombre de positions, on ne les compare pas.
        return None

    differences = []

    for x1, x2, in zip(position1, positions2):
        differences.append(abs(x1 - x2))

    return differences

In [161]:
print(
    compare_x_positions(
        block_x_positions[4],
        block_x_positions[5]
    )
)

[0.0, 0.0037078857421875, 0.005584716796875, 0.007476806640625]


In [ ]:
for i in range(len(block_x_positions) - 1): # Arrêt à l'avant dernier block pour comparaison
    current_block = block_x_positions[i]
    next_block = block_x_positions[i + 1]

    differences = compare_x_positions(
        current_block,
        next_block
    )

    print(i, "->", i + 1, ":", differences)

0 -> 1 : [0.7000007629394531, 18.599998474121094, 1069.1150131225586]
1 -> 2 : None
2 -> 3 : None
3 -> 4 : [4.657600402832031, 205.92059707641602, 306.5521125793457, 407.1836128234863]
4 -> 5 : [0.0, 0.0037078857421875, 0.005584716796875, 0.007476806640625]
5 -> 6 : [0.2539787292480469, 0.250274658203125, 0.248382568359375, 0.246490478515625]
6 -> 7 : [0.00848388671875, 0.00848388671875, 0.00848388671875, 0.00848388671875]
7 -> 8 : [0.2550048828125, 0.2550048828125, 0.2550048828125, 0.2550048828125]
8 -> 9 : [0.009510040283203125, 0.013214111328125, 0.015106201171875, 0.016998291015625]
9 -> 10 : [0.000995635986328125, 0.00469970703125, 0.006591796875, 0.00848388671875]
10 -> 11 : [0.000995635986328125, 0.00469970703125, 0.006591796875, 0.00848388671875]
11 -> 12 : [0.000995635986328125, 0.00469970703125, 0.006591796875, 0.00848388671875]
12 -> 13 : [0.000995635986328125, 0.00469970703125, 0.006591796875, 0.00848388671875]
13 -> 14 : [0.000995635986328125, 0.00469970703125, 0.006591796

In [163]:
differences = compare_x_positions(
    block_x_positions[4],
    block_x_positions[5]
)

mean_difference = sum(differences) / len(differences)

print(mean_difference)

0.004192352294921875


0.00419 est très petit. Celà confirme que les deux lignes ont quasiment la même structure horizontale.

In [165]:
mean_differences = []

for i in range(len(block_x_positions) - 1): # Arrêt à l'avant dernier block pour comparaison
    current_block = block_x_positions[i]
    next_block = block_x_positions[i + 1]

    differences = compare_x_positions(
        current_block,
        next_block
    )

    if differences is not None:
        mean_difference = sum(differences) / len(differences)
        mean_differences.append(mean_difference)

print(mean_differences)

[362.80500411987305, 231.07848072052002, 0.004192352294921875, 0.24978160858154297, 0.00848388671875, 0.2550048828125, 0.013707160949707031, 0.005192756652832031, 0.005192756652832031, 0.005192756652832031, 0.005192756652832031, 0.005192756652832031, 0.005192756652832031, 0.005192756652832031, 0.005192756652832031, 0.005192756652832031, 0.004192352294921875, 0.25, 0.0, 0.0041961669921875, 0.0052642822265625, 0.0052642822265625, 0.0052642822265625, 0.0052642822265625, 0.0052642822265625, 0.0052642822265625, 0.0052642822265625, 0.008544921875, 0.0138092041015625, 0.0, 0.0, 0.0, 0.0077056884765625, 0.0, 0.0052642822265625, 0.0052642822265625, 0.0052642822265625, 0.0041961669921875, 0.0052642822265625, 0.0052642822265625, 0.00897216796875, 0.00518798828125]


On voit deux ordres de grandeur très différents : 362 / 231 d’un côté, et autour de 0.004 à 0.25 de l’autre. Cette rupture explique le commencement du tableau.

Afin de chercher une rupture numérique, nous allons classer les valeurs du plus petit au plus grand car les valeus au dessu peuvent être mélangées.

In [166]:
sorted_mean_differences = sorted(mean_differences)

print(sorted_mean_differences)

[0.0, 0.0, 0.0, 0.0, 0.0, 0.004192352294921875, 0.004192352294921875, 0.0041961669921875, 0.0041961669921875, 0.00518798828125, 0.005192756652832031, 0.005192756652832031, 0.005192756652832031, 0.005192756652832031, 0.005192756652832031, 0.005192756652832031, 0.005192756652832031, 0.005192756652832031, 0.005192756652832031, 0.0052642822265625, 0.0052642822265625, 0.0052642822265625, 0.0052642822265625, 0.0052642822265625, 0.0052642822265625, 0.0052642822265625, 0.0052642822265625, 0.0052642822265625, 0.0052642822265625, 0.0052642822265625, 0.0052642822265625, 0.0077056884765625, 0.00848388671875, 0.008544921875, 0.00897216796875, 0.013707160949707031, 0.0138092041015625, 0.24978160858154297, 0.25, 0.2550048828125, 231.07848072052002, 362.80500411987305]


### Calcul des écarts entre deux valeurs triées consécutives

In [167]:
mean_gaps = []

for i in range(len(sorted_mean_differences) - 1):
    gap = (
        sorted_mean_differences[i + 1]
        - sorted_mean_differences[i]
    )

    mean_gaps.append(gap)

print(mean_gaps)

[0.0, 0.0, 0.0, 0.0, 0.004192352294921875, 0.0, 3.814697265625e-06, 0.0, 0.0009918212890625, 4.76837158203125e-06, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 7.152557373046875e-05, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00244140625, 0.0007781982421875, 6.103515625e-05, 0.00042724609375, 0.004734992980957031, 0.00010204315185546875, 0.23597240447998047, 0.00021839141845703125, 0.0050048828125, 230.82347583770752, 131.72652339935303]


In [168]:
print(len(sorted_mean_differences))
print(len(mean_gaps))

42
41


### Position du plus grand gap

In [170]:
max_gap = max(mean_gaps)

rupture_index = mean_gaps.index(max_gap)

print("Plus grand gap :", max_gap)
print("index :", rupture_index)

Plus grand gap : 230.82347583770752
index : 39


La rupture se trouve entre sorted_mean_differences[39] et sorted_mean_differences[40]

In [171]:
small_value = sorted_mean_differences[rupture_index]
large_value = sorted_mean_differences[rupture_index + 1]

print("Avant la rupture :", small_value)
print("Après la rupture :", large_value)

Avant la rupture : 0.2550048828125
Après la rupture : 231.07848072052002


### Seuil automatique comme auto_tolerance

In [172]:
similarity_threshold = (small_value + large_value) / 2

print("Seuil automatique :", similarity_threshold)

Seuil automatique : 115.66674280166626


### Comparaison des postions avec incorporation du seuil

In [175]:
for i in range(len(block_x_positions) - 1): # Arrêt à l'avant dernier block pour comparaison
    current_block = block_x_positions[i]
    next_block = block_x_positions[i + 1]

    differences = compare_x_positions(
        current_block,
        next_block
    )

    if differences is None:
        print(i, "->", i + 1, ": non comparable")
        continue

    mean_difference = sum(differences) / len(differences) 
    is_similar = mean_difference <= similarity_threshold

    print(i, "->", i + 1, ":", is_similar)

0 -> 1 : False
1 -> 2 : non comparable
2 -> 3 : non comparable
3 -> 4 : False
4 -> 5 : True
5 -> 6 : True
6 -> 7 : True
7 -> 8 : True
8 -> 9 : True
9 -> 10 : True
10 -> 11 : True
11 -> 12 : True
12 -> 13 : True
13 -> 14 : True
14 -> 15 : True
15 -> 16 : True
16 -> 17 : True
17 -> 18 : True
18 -> 19 : non comparable
19 -> 20 : non comparable
20 -> 21 : True
21 -> 22 : True
22 -> 23 : True
23 -> 24 : non comparable
24 -> 25 : non comparable
25 -> 26 : non comparable
26 -> 27 : non comparable
27 -> 28 : True
28 -> 29 : True
29 -> 30 : True
30 -> 31 : True
31 -> 32 : True
32 -> 33 : True
33 -> 34 : True
34 -> 35 : True
35 -> 36 : True
36 -> 37 : True
37 -> 38 : True
38 -> 39 : True
39 -> 40 : True
40 -> 41 : non comparable
41 -> 42 : non comparable
42 -> 43 : True
43 -> 44 : True
44 -> 45 : True
45 -> 46 : True
46 -> 47 : True
47 -> 48 : non comparable
48 -> 49 : non comparable
49 -> 50 : non comparable
50 -> 51 : non comparable
51 -> 52 : True
52 -> 53 : True
53 -> 54 : True
54 -> 55 : no

À partir de 4, plusieurs blocks consécutifs ont donc une géométrie horizontale similaire. Détection automatique d'une zone tabulaire.

Le soucis est que les valeurs manquantes dans le tableau peuvent donner "None" et donc nous perdons cet élément dans la stucture du tableau

1. Extraire les positions x des spans de chaque block
                    
2. Comparer les positions x de deux blocks consécutifs
                    
3. Calculer leur différence moyenne
                    
4. Comparer cette différence au seuil automatique
                    
5. True = structures similaires
   / False = structures différentes

### Mémorisation des blocks qui se ressemblent

In [180]:
similar_block_indices = []

for i in range(len(block_x_positions) - 1): # Arrêt à l'avant dernier block pour comparaison
    current_block = block_x_positions[i]
    next_block = block_x_positions[i + 1]

    differences = compare_x_positions(
        current_block,
        next_block
    )

    if differences is None:
        continue

    mean_difference = sum(differences) / len(differences) 

    if mean_difference <= similarity_threshold:
        similar_block_indices.append(i)
        similar_block_indices.append(i+1)

In [181]:
print(similar_block_indices)

[4, 5, 5, 6, 6, 7, 7, 8, 8, 9, 9, 10, 10, 11, 11, 12, 12, 13, 13, 14, 14, 15, 15, 16, 16, 17, 17, 18, 20, 21, 21, 22, 22, 23, 27, 28, 28, 29, 29, 30, 30, 31, 31, 32, 32, 33, 33, 34, 34, 35, 35, 36, 36, 37, 37, 38, 38, 39, 39, 40, 42, 43, 43, 44, 44, 45, 45, 46, 46, 47, 51, 52, 52, 53, 53, 54, 56, 57, 57, 58]


### Suppression des doublons

In [184]:
unique_indices = list(dict.fromkeys(similar_block_indices)) # dict.fromkeys() utilise chaque nombre comme clé d’un dictionnaire. Or un dictionnaire ne peut pas avoir deux fois la même clé. Donc ne garde qu'une seule clé sur plusieurs similaires

print(unique_indices)

[4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 20, 21, 22, 23, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 42, 43, 44, 45, 46, 47, 51, 52, 53, 54, 56, 57, 58]


### Transformation de la liste en groupe consécutifs

In [191]:
len(unique_indices)

46

In [ ]:
groups = [] # liste qui contiendra plus tard tous nos groupes

current_group = [unique_indices[0]] # premier indice

for index in unique_indices[1:]: # parcourt la liste à partir du 2ème élément
    if index == current_group[-1] +1: # si l'index de la boucle est égale au dernier élément de current_group + 1, alors il l'ajoute dans current_group
        current_group.append(index)

In [190]:
current_group

[4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18]

In [192]:
len(current_group)

15

Le premier bloc contient 15 blocks consécutifs des indices 4 à 18

In [ ]:
groups = [] # liste qui contiendra plus tard tous nos groupes

current_group = [unique_indices[0]] # premier indice

for index in unique_indices[1:]: # parcourt la liste à partir du 2ème élément
    if index == current_group[-1] +1: # si l'index de la boucle est égale au dernier élément de current_group + 1, alors il l'ajoute dans current_group
        current_group.append(index)

    else:                              # Lorsque current_group +1 différent d'index (ex: 18 -> 20), sauvegarde dans groups puis démarre current_group à 20
        groups.append(current_group)
        current_group = [index]

groups.append(current_group) # Pendant la boucle, groups.append(current_group) est déclenché lorsqu'on rencontre une rupture (18 -> 20)
                             # Mais pour le dernier groupe, il n'y a plus forcément de nombre après lui permettant de provoquer une rupture. La boucle se termine simplement.
                             # Donc on le sauvegarde manuellement une dernière fois.

In [197]:
print(groups)
print("Nombre de groupes sauvegardés :", len(groups))

[[4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18], [20, 21, 22, 23], [27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40], [42, 43, 44, 45, 46, 47], [51, 52, 53, 54], [56, 57, 58]]
Nombre de groupes sauvegardés : 6


On veut maintenant retrouver les vrais blocks PyMuPDF correspondants

### reconstruction d'une liste de blocks texte

In [198]:
text_blocks = []

for block in page_dict["blocks"]:
    if block["type"] == 0:
        text_blocks.append(block)

text_blocks[4] correspond au même block que block_x_positions[4]

## Test avec premier groupe

In [201]:
first_group = groups[0]

for block_index in first_group:
    block = text_blocks[block_index]

    texts = []

    for line in block["lines"]:
        for span in line["spans"]:
            text = span["text"].strip()

            if text:
                texts.append(text)

    print(block_index, "->", texts)

4 -> ['Afghanistan', '30 sept. 1966', '25 juin 1968', '25 juill. 1968']
5 -> ['Albanie', '15 oct. 1991', '15 oct. 1991', '14 nov. 1991']
6 -> ['Algérie', '17 avr. 1995', '21 fév. 1996', '22 mars 1996']
7 -> ['Angola', '14 juill. 2022', '21 sep. 2022', '21 oct. 2022']
8 -> ['Argentine', '21 mai 1991', '19 oct. 1994', '18 nov. 1994']
9 -> ['Arménie', '16 sep. 1992', '16 sep. 1992', '16 oct. 1992']
10 -> ['Australie', '24 mars 1975', '2 mai 1991', '1 juin 1991']
11 -> ['Autriche', '17 mai 1966', '25 mai 1971', '24 juin 1971']
12 -> ['Azerbaïdjan', '18 sep. 1992', '18 sep. 1992', '18 oct. 1992']
13 -> ['Bahamas', '19 oct. 1995', '19 oct. 1995', '18 nov. 1995']
14 -> ['Bahreïn', '22 sep. 1995', '14 fév. 1996', '15 mars 1996']
15 -> ['Bangladesh', '20 nov. 1979', '27 mars 1980', '26 avr. 1980']
16 -> ['Barbade', '13 mai 1981', '1 nov. 1983', '1 déc. 1983']
17 -> ['Bélarus', '10 juill. 1992', '10 juill. 1992', '9 août 1992']
18 -> ['Belgique', '15 déc. 1965', '27 août 1970', '26 sep. 1970']


### Affichage du premier block de chaque groupe

In [208]:
# Test
for i,j in enumerate(groups):
    print(i,j)

0 [4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18]
1 [20, 21, 22, 23]
2 [27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40]
3 [42, 43, 44, 45, 46, 47]
4 [51, 52, 53, 54]
5 [56, 57, 58]


In [209]:
for group_index, group in enumerate(groups):
    first_block_index = group[0]
    block = text_blocks[first_block_index]

    texts = []

    for line in block["lines"]:
        for span in line["spans"]:
            text = span["text"].strip()

            if text:
                texts.append(text)

    print(group_index, "->", first_block_index, "->", texts)

0 -> 4 -> ['Afghanistan', '30 sept. 1966', '25 juin 1968', '25 juill. 1968']
1 -> 20 -> ['Bénin', '10 sep. 1965', '6 sep. 1966', '14 oct. 1966']
2 -> 27 -> ['Botswana', '15 janv. 1970', '15 janv. 1970', '14 fév. 1970']
3 -> 42 -> ['Danemark', '11 oct. 1965', '24 avr. 1968', '24 mai 1968']
4 -> 51 -> ['Congo, République démocratique du', '29 oct. 1968', '29 avr. 1970', '29 mai 1970']
5 -> 56 -> ['Espagne', '21 mars 1994', '18 août 1994', '17 sep. 1994']


In [210]:
block_index = first_group[0]
block = text_blocks[block_index]

print("Block bbox :", block["bbox"])

for line in block["lines"]:
    for span in line["spans"]:
        print(
            span["text"],
            "->",
            span["bbox"]
        )

Block bbox : (40.95759963989258, 298.1300964355469, 491.42364501953125, 309.256591796875)
Afghanistan -> (40.95759963989258, 298.1300964355469, 88.77861022949219, 309.256591796875)
30 sept. 1966 -> (242.22059631347656, 298.2575988769531, 294.3757629394531, 308.71258544921875)
25 juin 1968 -> (342.85211181640625, 298.2575988769531, 389.39813232421875, 308.71258544921875)
25 juill. 1968 -> (443.4836120605469, 298.2575988769531, 491.42364501953125, 308.71258544921875)


Une ligne de tableau avec deux niveaux de provenance, le block entier puis à l'intérieur chaque cellule qui possède sa propre bbox

### Construction d'un objet structuré

In [217]:
row_data = {
    "page": 34, # index (35ème page)
    "bbox": block["bbox"], # position de la ligne entière
    "cells": [] # liste qui contiendra nos cellules
}

### Remplissage de cell

In [218]:
for line in block["lines"]:
    for span in line["spans"]:
        text = span["text"].strip()

        if text:
            row_data["cells"].append({
                "text": text,
                "bbox": span["bbox"]
            })

In [219]:
print(row_data)

{'page': 34, 'bbox': (40.95759963989258, 298.1300964355469, 491.42364501953125, 309.256591796875), 'cells': [{'text': 'Afghanistan', 'bbox': (40.95759963989258, 298.1300964355469, 88.77861022949219, 309.256591796875)}, {'text': '30 sept. 1966', 'bbox': (242.22059631347656, 298.2575988769531, 294.3757629394531, 308.71258544921875)}, {'text': '25 juin 1968', 'bbox': (342.85211181640625, 298.2575988769531, 389.39813232421875, 308.71258544921875)}, {'text': '25 juill. 1968', 'bbox': (443.4836120605469, 298.2575988769531, 491.42364501953125, 308.71258544921875)}]}


row_data

│
├── page

├── bbox de toute la ligne

│

└── cells

    ├── cellule 0 : texte + bbox

    ├── cellule 1 : texte + bbox

    ├── cellule 2 : texte + bbox
    
    └── cellule 3 : texte + bbox

In [223]:
print(row_data["cells"][0]["text"])
print(row_data["cells"][2]["text"])
print(row_data["cells"][2]["bbox"])

Afghanistan
25 juin 1968
(342.85211181640625, 298.2575988769531, 389.39813232421875, 308.71258544921875)


### Objet structuré pour tout first_group

In [226]:
first_group

[4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18]

In [232]:
table_rows = []

for block_index in first_group:
    block = text_blocks[block_index]

    row_data = {
        "page": 34,
        "bbox": block["bbox"],
        "cells": []
    }

    for line in block["lines"]:
        for span in line["spans"]:
            text = span["text"].strip()

            if text:
                row_data["cells"].append({
                    "text": text,
                    "bbox": span["bbox"]
                })

    table_rows.append(row_data)

In [234]:
for row in table_rows[:2]:
    print(row)
    print("-" * 80)

{'page': 34, 'bbox': (40.95759963989258, 298.1300964355469, 491.42364501953125, 309.256591796875), 'cells': [{'text': 'Afghanistan', 'bbox': (40.95759963989258, 298.1300964355469, 88.77861022949219, 309.256591796875)}, {'text': '30 sept. 1966', 'bbox': (242.22059631347656, 298.2575988769531, 294.3757629394531, 308.71258544921875)}, {'text': '25 juin 1968', 'bbox': (342.85211181640625, 298.2575988769531, 389.39813232421875, 308.71258544921875)}, {'text': '25 juill. 1968', 'bbox': (443.4836120605469, 298.2575988769531, 491.42364501953125, 308.71258544921875)}]}
--------------------------------------------------------------------------------
{'page': 34, 'bbox': (40.95759963989258, 318.09320068359375, 488.8291931152344, 329.2196960449219), 'cells': [{'text': 'Albanie', 'bbox': (40.95759963989258, 318.09320068359375, 69.85759735107422, 329.2196960449219)}, {'text': '15 oct. 1991', 'bbox': (242.22430419921875, 318.220703125, 286.7983093261719, 328.6756896972656)}, {'text': '15 oct. 1991', '

In [252]:
block = text_blocks[first_group[0]]

for line in block["lines"]:
    for span in line["spans"]:
        x0 = span["bbox"][0]

        column_index = min(
            range(len(column_positions)),
            key=lambda i: abs(x0 - column_positions[i])
        )

        print(span["text"], "-> colonne", column_index)

Afghanistan -> colonne 0
30 sept. 1966 -> colonne 1
25 juin 1968 -> colonne 2
25 juill. 1968 -> colonne 3


In [ ]:
table_rows = []

for block_index in first_group:
    block = text_blocks[block_index]

    row_data = {
        "page": 34,
        "bbox": block["bbox"],
        "cells": []
    }

    for line in block["lines"]:
        for span in line["spans"]:
            text = span["text"].strip()

            if text:
                x0 = span["bbox"][0] # Pour chaque texte de cette ligne, on récupère sa position horizontale

                column_index = min(                     # On cherche la colonne dont la position moyenne est la plus proche
                    range(len(column_positions)),       # column_positions = [41.01, 242.27, 342.91, 443.54] donc 4 positions
                    key=lambda i: abs(x0 - column_positions[i]) # Pour un indice i, calcul de la distance entre le x0 du texte et la position de la colonne i
                )

                row_data["cells"].append({
                    "column_index": column_index,
                    "text": text,
                    "bbox": span["bbox"]
                })

    table_rows.append(row_data)

In [254]:
for row in table_rows[:2]:
    for cell in row["cells"]:
        print(cell["column_index"], "->", cell["text"])

    print("-" * 50)

0 -> Afghanistan
1 -> 30 sept. 1966
2 -> 25 juin 1968
3 -> 25 juill. 1968
--------------------------------------------------
0 -> Albanie
1 -> 15 oct. 1991
2 -> 15 oct. 1991
3 -> 14 nov. 1991
--------------------------------------------------


### Recap

PDF

 │
 
1. Extraction des blocs/spans avec PyMuPDF

 │
 
2. Repérer les blocs qui ressemblent à des lignes de tableau

 │   → similarity_threshold
 
3. Former un groupe candidat de lignes de tableau

 │   → first_group
 
4. Observer les positions X dans ce groupe

 
 │   → auto_tolerance
 
5. Regrouper les X proches pour découvrir les colonnes

 │   → column_positions
 
6. Affecter chaque span à la colonne la plus proche

 │
 
7. Construire table_rows avec text + column_index + bbox

similarity_threshold

        |

"Quelles LIGNES vont ensemble ?"
        
        |

     first_group

#################################################

auto_tolerance

        |

"Quelles positions X vont ensemble ?"

        |
        
  column_positions

# Création du code étape par étape depuis le début

In [255]:
doc = pymupdf.open(pdf_path)

print(f"Nombre de pages : {len(doc)}")
print(f"Métadonnées : {doc.metadata}")

Nombre de pages : 38
Métadonnées : {'format': 'PDF 1.4', 'title': '', 'author': '', 'subject': '', 'keywords': '', 'creator': 'Adobe InDesign 20.5 (Macintosh)', 'producer': 'Adobe PDF Library 17.0', 'creationDate': "D:20250908153744-06'00'", 'modDate': "D:20250924155840-04'00'", 'trapped': '', 'encryption': None}


In [256]:
for page_index in range(len(doc)):
    page = doc[page_index]
    page_dict = page.get_text("dict")

    print("Page :", page_index)

Page : 0
Page : 1
Page : 2
Page : 3
Page : 4
Page : 5
Page : 6
Page : 7
Page : 8
Page : 9
Page : 10
Page : 11
Page : 12
Page : 13
Page : 14
Page : 15
Page : 16
Page : 17
Page : 18
Page : 19
Page : 20
Page : 21
Page : 22
Page : 23
Page : 24
Page : 25
Page : 26
Page : 27
Page : 28
Page : 29
Page : 30
Page : 31
Page : 32
Page : 33
Page : 34
Page : 35
Page : 36
Page : 37


À l'intérieur de chaque page, on va progressivement remettre:

1. récupérer les blocs texte

2. récupérer leurs positions X

3. comparer les blocs
        ->
   similarity_threshold

4. créer les groupes de blocs similaires
        ->
   groupes candidats de tableaux

5. pour chaque groupe candidat :
   récupérer ses positions X
        ->
   auto_tolerance
        ->
   column_positions

6. attribuer les spans aux colonnes
        ->
   column_index

7. construire les lignes
        ->
   table_rows

8. conserver :
   page + bbox + text + column_index